### PR TITLE
Make AOV::write_image() virtual

### PR DESCRIPTION
**Pull request by usakhelo**
Hi @est77,
This is quick version of the change that you've suggested me to do. I just moved some writing functionality from frame.cpp into AOV class and made the function virtual.
I tested it on max scene and it worked.
If I understand correctly this change will not reduce cryptomatte's write_image code because of restriction of `GenericImageFileWriter` to 4 channels. Cryptomatte will still need its own FileWriter.

### CODE REVIEW

**Comment by est77** [comment]
Yes, this is the idea. The Cryptomatte AOV will override write_image() and use its own writer.

File: src/appleseed/renderer/modeling/frame/frame.cpp [hunk 1]
```diff
@@ -717,8 +705,8 @@ bool Frame::write_aov_images(const char* file_path) const
         const string aov_file_path = (directory / aov_file_name).string();
 
         // Write AOV image.
-        if (!write_image(aov_file_path.c_str(), aov->get_image(), aov))
-            success = false;
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/frame/frame.cpp]: Why the change of logic here?

File: src/appleseed/renderer/modeling/frame/frame.cpp [hunk 2]
```diff
@@ -763,8 +751,9 @@ bool Frame::write_main_and_aov_images() const
                 filepath = new_filepath;
             }
 
-            if (!write_image(filepath.string().c_str(), aov->get_image(), aov))
-                success = false;
+            // Write AOV image.
+            if (!aov->write_images(filepath.string().c_str()))
+                return false;
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/frame/frame.cpp]: And here?

File: src/appleseed/renderer/modeling/aov/aov.cpp [hunk 3]
```diff
@@ -94,6 +99,37 @@ void AOV::create_image(
...
 
+bool AOV::write_images(const char* file_path) const
+{
+    try
+    {
+        GenericImageFileWriter writer(file_path);
+
+        writer.append_image(&get_image());
+
+        if(has_color_data())
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/aov/aov.cpp]: Insert space after `if`.

File: src/appleseed/renderer/modeling/aov/aov.cpp [hunk 4]
```diff
@@ -94,6 +99,37 @@ void AOV::create_image(
...
+            writer.set_image_output_format(PixelFormatHalf);
+
+        writer.set_image_channels(get_channel_count(), get_channel_names());
+
+        ImageAttributes image_attributes = ImageAttributes::create_default_attributes();
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/aov/aov.cpp]: Include `foundation/image/imageattributes.h`.

File: src/appleseed/renderer/modeling/aov/aov.h [hunk 5]
```diff
@@ -92,6 +92,9 @@ class APPLESEED_DLLSYMBOL AOV
     // Apply any post-processing needed to the AOV image.
     virtual void post_process_image(const Frame& frame);
 
+    // Write image to EXR file.
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/aov/aov.h]: *OpenEXR*

Comment (dictoon) [reply src/appleseed/renderer/modeling/aov/aov.h]: You called the method `write_images()` (plural), is this intended?

File: src/appleseed/renderer/modeling/frame/frame.cpp (hunk 1 above)
Comment (usakhelo) [reply src/appleseed/renderer/modeling/frame/frame.cpp]: You are asking about `success` variable, right? Sorry I misunderstood. It's because of the for loop. I thought breaking the loop and returning false on first failing aov.

File: src/appleseed/renderer/modeling/aov/aov.h (hunk 5 above)
Comment (usakhelo) [reply src/appleseed/renderer/modeling/aov/aov.h]: I think standalone method in `frame.cpp` was called plural, and I made copy of that.

File: src/appleseed/renderer/modeling/frame/frame.cpp (hunk 1 above)
Comment (dictoon) [reply src/appleseed/renderer/modeling/frame/frame.cpp]: Yes, the `success` variable. I prefer that we don't change the logic, if possible.

File: src/appleseed/renderer/modeling/aov/aov.h (hunk 5 above)
Comment (dictoon) [reply src/appleseed/renderer/modeling/aov/aov.h]: Ok.
